cr-db/cr/db/loader.py

```diff
@@ -1,53 +1,158 @@
 
-    with file(filename) as the_file:
-        objs = json.load(the_file)
-        for obj in objs:
-            collection.insert(obj)
+        collection = getattr(db, obj_name)
 
+        if clear:
+            collection.remove()
 
-def load_dataset(csv_filename, db):
-    with file(csv_filename, 'rU') as csv_file:
-        csv_data = csv.reader(csv_file)
-        headers = csv_data.next()
+        with file(filename) as the_file:
+            if optimize_for == 'speed':
+                objs = json.load(the_file)
+                collection.insert(objs)
+            elif optimize_for == 'memory':
+                objs = ijson.items(the_file, 'item')
+                for obj in objs:
+                    collection.insert(obj)
 
-        last_header = None
-        for i, header in enumerate(headers):
-            if header:
-                last_header = header
-            else:
-                # multiple response have no header
-                headers[i] = last_header
+    @classmethod
+    def load_dataset(cls, csv_filename, db, modify_to, max_categories):
+        '''Convert and load matrix-like data from CSV to MongoDB.'''
+        with file(csv_filename, 'rU') as csv_file:
+            csv_data = csv.reader(csv_file)
+            headers = csv_data.next()
 
-        columns = [[] for _ in headers]
-        for r, row in enumerate(csv_data):
-            for i, data in enumerate(row):
-                if data:
-                    columns[i].append(str(data))
+            last_header = None
+            for i, header in enumerate(headers):
+                if header:
+                    last_header = header
                 else:
-                    columns[i].append(None)
+                    # multiple response have no header
+                    headers[i] = last_header
+
+            columns = [[] for _ in headers]
+            for _, row in enumerate(csv_data):
+                for i, data in enumerate(row):
+                    if data:
+                        columns[i].append(str(data))
+                    else:
+                        columns[i].append(None)
 
-        data = {'headers': headers,
+            encoding = []
+            decoding = []
+            for i, column in enumerate(columns):
+                modified, encode, decode = cls._modify_column(
+                    column, modify_to, max_categories
+                )
+                columns[i] = modified
+                encoding.append(encode)
+                decoding.append(decode)
+
+            data = {
+                'headers': headers,
                 'columns': columns,
-                }
+                'encoding': encoding,
+                'decoding': decoding,
+                'modified_to': modify_to,
+            }
+
+            return db.datasets.insert(data, check_keys=False)
+
+    @classmethod
+    def select_filtered_data(cls, dataset, filter_col, filter_val, group_by,
```

Comment:
@percious I'm not sure what you mean by 'persist'. The filter is passed in as a function parameter, which allows it to take multiple values in different scenarios. In order to generate the requested data, we have the following steps:
1. Selection
2. Filtering
3. Grouping
4. Preprocessing
And they need to be in that particular order, since some of the groups may be filtered out in the 2nd step.

Anyhow, I've factored out the category filtering, for a particular column, into a separate function. This allows for better reusability. Also, the preparation of the data is factored out into a separate function (packaging into numpy array).

Please let me know if you had something else in mind...